---
title: "Beghini_2019_16S_smoking"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    echo = TRUE,
    message = FALSE,
    warning = FALSE
)
```

```{r load packages}
library(benchdamic)
library(phyloseq)
library(MicrobiomeBenchmarkData)
library(dplyr)
```

```{r import dataset}
dat_names <- getDataset()
tse <- getDataset(dat_names[1], dryrun = FALSE)[[1]]
```

```{r}
fname <- "https://raw.githubusercontent.com/waldronlab/nychanesmicrobiome/master/inst/extdata/biosis.tsv"
biosis <- readr::read_tsv(fname, col_names = TRUE) %>% 
    magrittr::set_colnames(c("GENUS", "BIOSIS"))
```


```{r}
rowData(tse) <- tse %>% 
    rowData() %>% 
    as.data.frame() %>% 
    tibble::rownames_to_column(var = "OTU") %>% 
    mutate(
        Genus = sub("^.+__", "", Genus),
        Genus = sub(" .*$", "", Genus),
    ) %>% 
    rename(GENUS = Genus) %>% 
    left_join(biosis, by = "GENUS") %>% 
    tibble::column_to_rownames(var = "OTU") %>% 
    as.data.frame() %>% 
    DataFrame()
```

Create prior knowledge info

```{r prior knowledge}
biologicalInfo <- tse %>% 
    rowData() %>% 
    as.data.frame() %>% 
    select(GENUS, BIOSIS) %>% 
    mutate(
        newNames = paste0(rownames(.), "|", GENUS),
        BIOSIS = ifelse(is.na(BIOSIS), "unknown", BIOSIS)
    ) %>% 
    rename(Type = BIOSIS)
```

Convert to phyloseq

```{r convert to phyloseq}
## Biosis information is lost, but it's already stored in the biologicalInfo
## variable
ps <- mia::makePhyloseqFromTreeSummarizedExperiment(tse)
```

Useful variables

```{r useful variables}
grp <- "smokingstatus"
contrast <- c("never_smoker", "cigarette")
```

Set reference

```{r set reference}
sample_data(ps)[[grp]] <- factor(sample_data(ps)[[grp]])
sample_data(ps)[[grp]] <- relevel(
    x = sample_data(ps)[[grp]], ref = contrast[1]
)
```

Prepare normalization

```{r normalization}
## Normalization methods
norm_pars <- tibble::tribble(
    ~fun, ~method,
    "norm_edgeR", "none",
    "norm_edgeR", "TMM",
    "norm_DESeq2", "poscounts",
    "norm_CSS", "median"
)

## set normalization
my_norm <- setNormalizations(fun = norm_pars$fun, method = norm_pars$method)

## Run normalization
ps <- runNormalizations(normalization_list = my_norm, object = ps)
```

Create weights

```{r create weights}
zinbweights <- weights_ZINB(
    object = ps,
    K = 0,
    design = "~ 1"
)

```

Prepare DA methods

```{r prepare DA methods}
# edgeR
my_edger <- set_edgeR(
    group_name = grp,
    design = as.formula(paste0("~", grp)),
    norm = "TMM",
    coef = 2
)

# DESeq2
my_deseq2 <- set_DESeq2(
    contrast = c(grp, contrast),
    design = as.formula(paste0("~", grp)),
    norm = "poscounts"
)

# limma 
my_limma <- set_limma( # I get a warning
    design = as.formula(paste0("~", grp)),
    norm = c("TMM", "CSSmedian"),
    coef = 2
)

# metagenomeSeq
my_metagenomeseq <- set_metagenomeSeq(
    design = as.formula(paste0("~", grp)),
    norm = "CSSmedian",
    coef = 2
)

# ALDEx2
my_aldex2 <- set_ALDEx2(
    conditions = grp,
    test = "t",
    norm = "none"
)

# corncob
my_corncob <- set_corncob(
    formula = as.formula(paste0("~", grp)),
    phi.formula = as.formula(paste0("~", grp)),
    formula_null = ~ 1,
    phi.formula_null = as.formula(paste0("~", grp)),
    test = "Wald",
    coefficient = paste0(grp, contrast[2]),
    norm = "none"
)

# MAST
my_mast <- set_MAST(
    rescale = "median",
    design = as.formula(paste0("~", grp)),
    coefficient = paste0(grp, contrast[2]),
    norm = "none"
)

# Seurat
my_seurat <- set_Seurat(
    test.use = "wilcox",
    contrast = c(grp, contrast),
    norm = "none"
)

my_methods <- c(
    my_edger, my_deseq2, my_limma, my_metagenomeseq, my_aldex2,
    my_corncob, my_mast, my_seurat
)
```

Prepare directions (for enrichment)

```{r direction}
direction <- c(
    edgeR.TMM = "logFC", 
    DESeq2.poscounts = "log2FoldChange",
    limma.CSSmedian = "logFC",
    limma.TMM = "logFC",
    metgenomeSeq.CSSmedian = paste0(grp, contrast[2]),
    ALDEx2.none = "effect",
    corncob.none = "Estimate",
    MAST.none = "logFC",
    Seurat.none = "avg_log2FC"
)
```

# Run DA

```{r run DA methods}
DA <- runDA(method_list = my_methods, object = ps, weights = zinbweights)
```

## Run ANCOMBC

```{r}
ancombc <- function(ps, formula, group) {
    out <- ANCOMBC::ancombc(phyloseq = ps, formula = formula, 
        p_adj_method = "bonferroni", zero_cut = 0.90, lib_cut = 1000, 
        group = group, struc_zero = TRUE, neg_lb = TRUE, 
        tol = 1e-5, max_iter = 100, conserve = TRUE, alpha = 0.05, 
        global = TRUE)
    res <- out$res
    ### extract important statistics ###
    vector_of_pval <- res$p_val[[1]] # contains the p-values
    vector_of_adjusted_pval <- res$q_val[[1]] # contains the adjusted p-values
    name_of_your_features <- rownames(res$p_val) # contains the OTU, or ASV, or other feature 
    # names. Usually extracted from the rownames of 
    # the count data
    vector_of_logFC <- res$beta[[1]] # logos the logFCs
    vector_of_statistics <- res$beta[[1]] # contains other statistics
    
    ### prepare the output ###
    pValMat <- data.frame("rawP" = vector_of_pval,
        "adjP" = vector_of_adjusted_pval)
    statInfo <- data.frame("logFC" = vector_of_logFC,
        "statistics" = vector_of_statistics) 
    name <- "ANCOMBC"
    # Be sure that your method hasn't changed the order of the features. If it 
    # happens, you'll need to re-establish the original order.
    rownames(pValMat) <- rownames(statInfo) <- name_of_your_features 
    
    # Return the output as a list
    return(list("pValMat" = pValMat, "statInfo" = statInfo, "name" = name))
}
my_ancombc <-  ancombc(ps, grp, grp)
DA$ancombc <- my_ancombc
```

# Enrichment

## Run enrichment

```{r create enrichment}
enrichment <- createEnrichment(
    object = DA,
    priorKnowledge = biologicalInfo,
    enrichmentCol = "Type",
    namesCol = "newNames",
    slot = "pValMat",
    colName = "adjP",
    type = "pvalue",
    direction = direction,
    threshold_pvalue = 0.1,
    threshold_logfc = 0,
    top = NULL,
    alternative = "greater",
    verbose = TRUE
)
```

## Plot enrichment

```{r, fig.height=7}
plotEnrichment(
    enrichment = enrichment, 
    enrichmentCol = "Type",
    levels_to_plot = c("Aerobic", "Anaerobic")
)

```

# TRUE and FALSE positives

## Create table of positives

```{r create positives}
positives <- createPositives(
    object = DA, 
    priorKnowledge = biologicalInfo,
    enrichmentCol = "Type",
    namesCol = "newNames",
    slot = "pValMat",
    colName = "rawP",
    type = "pvalue",
    direction = direction,
    threshold_pvalue = 1,
    threshold_logfc = 0,
    top = seq.int(from = 0, to = 50, by = 5),
    alternative = "greater",
    verbose = FALSE,
    TP = list(c("DOWN Abundant", "Anaerobic"), c("UP Abundant", "Aerobic")),
    FP = list(c("DOWN Abundant", "Aerobic"), c("UP Abundant", "Anaerobic"))
)
```

Table of positives:

```{r head positives}
head(positives)
```


## Plot positives

```{r}
plotPositives(positives)
```

