---
title: "HMP_2012_16S_gingival_V35_subset"
output:
  pdf_document:
    latex_engine: xelatex
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, message=FALSE}
library(MicrobiomeBenchmarkData)
library(magrittr) # must be attached for the functions in functions_da.R to work
library(dplyr)
library(purrr)
## Project directory must be selected in the knitr menu
source("R/functions_da.R") 
```

```{r}
tse <- getDataset("HMP_2012_16S_gingival_V35_subset", dryrun = FALSE)[[1]]
## A couple of useful variables
ref <- "subgingival_plaque"
grp <- "hmp_body_subsite"
```

## Differential abundance analysis

Trying a few methods for differential abundance analysis

```{r compute DA, message=FALSE}
DA_filename <- "rmarkdown/HMP_2012_16S_ginvival_V35_subset_DA.rds"

if (file.exists(DA_filename)) {
  
    cat(paste0("File ", DA_filename, " already exists. Loading it..."))
  
    DA <- readRDS(DA_filename)
    
} else {
    
    cat(paste0("File ", DA_filename, " does not exist. Calculating DA..."))
  
    dds <-  deseq2Poscounts(tse, grp, ref)
    dge <- edgerTmm(tse, grp, ref)
    lv <- limmaVoomTmm(tse, grp, ref)
    al2 <- aldex2(tse, grp, ref)
    mgs <- metagenomeSeq(tse, grp, ref)
    cc <- corncob(tse, grp, ref, fdr_cutoff = 0.1)
    ambc <- ancom_bc(tse, grp, ref)
    
    DA <- list(
        deseq2_poscounts = dds, edgeR_tmm = dge, limma_voom = lv, 
        aldex2 = al2, metagenomeseq = mgs, corncob = cc, 
        ancom_bc = ambc
    )
    
    saveRDS(DA, "rmarkdown/HMP_2012_16S_ginvival_V35_subset_DA.rds")
}

```

## Enrichment


## Putatite  True Positives - Putative False Positives



