---
title: "Lefser"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r}
library(MicrobiomeBenchmarkData)
library(lefser)
library(curatedMetagenomicData)
```

Define a centered log ratio transformation function:

```{r}
clr <- function(x) {
    ## Centered log ratio transformation of a vector
    ## Sources: 
    ## + https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6755255/
    ## + https://www.r-bloggers.com/2021/08/calculate-geometric-mean-in-r/
    if (any(is.na(x) | any(x <= 0)))
        stop("Input vector must not contain NAs, 0s, or negative numbers.")
    log(x / exp(mean(log(x))))
}
```



```{r}
tse <- getDataset("HMP_2012_16S_gingival_V35_subset", dryrun = FALSE)[[1]]
se <- SummarizedExperiment::SummarizedExperiment(
    assays = list(counts = assay(tse)),
    colData = colData(tse),
    rowData = rowData(tse)
)
se
```

## Data from MicrobiomeBenchmarkData

```{r}
allRanks <- mia::splitByRanks(se)
se_genus <- allRanks$GENUS
assays(se_genus)[["clr"]] <- se_genus |>
    assay() |>
    t() |>
    {\(y) Hotelling::clr(y + 1)}() |>
    t() 
se_genus
```

Direct counts:

```{r}
res_genus_counts <- lefser(se_genus, groupCol = "hmp_body_subsite", assay = 1L)
res_genus_counts
```

CLR transform:

```{r}
res_genus_clr <- lefser(se_genus, groupCol = "hmp_body_subsite", assay = 2L)
dim(res_genus_clr)
```

## With the ZellerG_2014 dataset

### Imported from curatedMetagenomicData v3.2.3

```{r}
zeller_tse <- curatedMetagenomicData(
    "ZellerG_2014.relative_abundance", counts = TRUE, rownames = "short",
    dryrun = FALSE
)[[1]]

zeller_se <- SummarizedExperiment(
    assays = list(counts = assay(zeller_tse)),
    colData = colData(zeller_tse),
    rowData = rowData(zeller_tse)
)

zeller <- zeller_se[,zeller_se$study_condition != "adenoma"]
table(zeller$study_condition)
```

Using the counts directly

```{r}
res <- lefser(zeller, groupCol = "study_condition", assay = 1)
dim(res)
```


```{r}
res_age <- lefser(
    zeller, groupCol = "study_condition", blockCol = "age_category", assay = 1
)
dim(res_age)
```


Using clr transform

```{r}
assays(zeller)[["clr"]] <- zeller |>
    assay() |>
    {\(y) clr(y + 1)}()
```




```{r}
res_clr <- lefser(zeller, groupCol = "study_condition", assay = 2)
dim(res_clr)
```

```{r}
res_clr_age <- lefser(
    zeller, groupCol = "study_condition", blockCol = "age_category", assay = 2
)
dim(res_clr_age)
```

## Data included in lefser

```{r}
data("zeller14")
zeller14 <- zeller14[,zeller14$study_condition != "adenoma"]
assays(zeller14)[["clr"]] <- zeller14 |> 
    assay() |> 
    {\(y) apply(y + 1, 2, clr)}()
zeller14    
```

```{r}
res <- lefser(
    zeller14, groupCol = "study_condition", blockCol = "age_category",
    assay = 1
)
dim(res)
```
```{r}
lefserPlot(res)
```


```{r}
res_clr <- lefser(
    zeller14, groupCol = "study_condition", blockCol = "age_category",
    assay = 2
)
dim(res_clr)
```

