---
title: "Ravel_2013_16S_BV"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    echo = TRUE,
    message = FALSE
)
```


This is a study in which the daily temporal dynamics of the vaginal microbiome
were recorded and reported for 25 women during a period of 10 weeks. Fifteen, 
six, and four of those women were diagnosed with sympomatic bacterial vaginosis
(SBV), asymptomatic bacterial vaginosis (BV), or were never diagnosed with 
BV, respectively.

```{r}
library(SummarizedExperiment)
library(purrr)
library(dplyr)
```


Download the datasets

```{r}
## Sample metadata
url1 <- "https://static-content.springer.com/esm/art%3A10.1186%2F2049-2618-1-29/MediaObjects/40168_2013_28_MOESM1_ESM.xlsx"
metadata_file <- tempfile()
download.file(url = url1, destfile = metadata_file)

## OTUs
url2 <- "https://static-content.springer.com/esm/art%3A10.1186%2F2049-2618-1-29/MediaObjects/40168_2013_28_MOESM3_ESM.xlsx"
otu_table_file <- tempfile()
download.file(url = url2, destfile = otu_table_file)

```

Import datasets

```{r}
## Import sample metadata
metadata <- readxl::read_xlsx(
    metadata_file, range = "Metadata!A4:T1760"
)

## Import taxa table
otu_table <- readxl::read_xlsx(
    otu_table_file, range = "A4:EW1661"
)

## A few changes needed
sample_metadata <- metadata %>% 
    left_join(otu_table[c("sampleID", "Total Reads")], by = "sampleID") %>% 
    tibble::column_to_rownames(var = "sampleID") %>% 
    as.data.frame()

relative_abundance_matrix <- otu_table %>% 
    select(-`Total Reads`) %>% 
    tibble::column_to_rownames(var = "sampleID") %>% 
    as.data.frame() %>% 
    as.matrix() %>% 
    t()
```

```{r}
dim(sample_metadata)
dim(relative_abundance_matrix)
```

```{r}
interesect_samples <- 
    intersect(colnames(relative_abundance_matrix), rownames(sample_metadata))
```

```{r}
se <- SummarizedExperiment(
    assays = S4Vectors::SimpleList(
        relative_abundance = relative_abundance_matrix[, interesect_samples]
    ),
    colData = S4Vectors::DataFrame(sample_metadata[interesect_samples,])
)
```

```{r}
assays(se) <- S4Vectors::SimpleList(
    counts = t(t(assay(se)) * colData(se)$Total.Reads / 100),
    relative_abundance = relative_abundance_matrix[, interesect_samples]
)
```






```{r, eval=FALSE}
## Several steps were necessary to get the taxonomies--for some reason taxize
## kept asking for an NCBI API key even when there were only a few taxa.
## So, I opted for saving the taxonomies in a file.

# taxa <- sub(" .+$", "", rownames(se))
# taxonomies1 <- taxize::classification(taxa[1:32], db = "ncbi")
# taxonomies2 <- taxize::classification(taxa[33], db = "ncbi")
# taxonomies3 <- taxize::classification(taxa[34:39], db = "ncbi")
# taxonomies4 <- taxize::classification(taxa[40], db = "ncbi")
# taxonomies5 <- taxize::classification(taxa[41:151], db = "ncbi")
# taxonomies <- c(
#     taxonomies1, taxonomies2, taxonomies3, taxonomies4, taxonomies5
# )
# 
# rownames(se)[which(is.na(taxonomies))]
# taxonomies[[grep("candidate", names(taxonomies))]] <- 
#     taxize::classification("candidate division TM7", db = "ncbi")[[1]]
# names(taxonomies)[grep("candidate", names(taxonomies))] <- "candidate division TM7"
# 
# saveRDS(taxonomies, "Ravel_2013_BV_taxonomies.rds")

```


```{r, eval=FALSE}
# taxonomies <- readRDS(file = "Ravel_2013_BV_taxonomies.rds")
```

```{r, eval=FALSE}
# taxToTable <- function(tax) {
#     
#     valid_ranks <- c(
#     "superkingdom", "kingdom", "phylum", "class", "order", "family", "genus"
#     )
#     
#     if (sum(is.na(tax)) == 1) {
#         n <- length(valid_ranks)
#         df <- 
#             matrix(rep(NA, n), ncol = n, dimnames = list(NULL, valid_ranks)) |>
#             tibble::as_tibble()
#         return(df)
#     }
#     
#     valid_ranks <- c(
#     "superkingdom", "kingdom", "phylum", "class", "order", "family", "genus"
#     )
#     x <- tax %>% 
#         dplyr::filter(.data[["rank"]] %in% .env[["valid_ranks"]]) |>
#         dplyr::select(name, rank)
#     df <- matrix(x$name, ncol = length(x$name), dimnames = list(NULL, x$rank)) |>
#     tibble::as_tibble()
#     return(df)
# }
```

```{r, eval=FALSE}
# taxonomy_table <- map(taxonomies, ~ taxToTable(.x)) %>% 
#     bind_rows() %>%
#     discard(~all(is.na(.x))) %>% 
#     as.data.frame()
# rownames(taxonomy_table) <- rownames(se)
# write.table(taxonomy_table, "Ravel_2013_BV_taxonomy_table.tsv", sep = "\t", 
#     row.names = TRUE, col.names = TRUE)
```

```{r}
taxonomy_table <- 
    read.table("Ravel_2013_16S_BV_taxonomy_table.tsv", sep = "\t", header = TRUE, row.names = 1)
```

```{r}
rowData(se) <- S4Vectors::DataFrame(taxonomy_table)
```


```{r}
se
```
```{r}
saveRDS(se, "Ravel_2013_BV_se.rds")
```

```{r}
# colData(se) %>% 
#     as.data.frame() %>% 
#     filter(SBV == 1 | ABV == 1) %>% 
#     select(subjectID)
```

