---
title: "Ravel_2013_BV"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE)
```

```{r}
library(SummarizedExperiment)
library(dplyr)
library(tidyr)
library(tibble)
library(purrr)
library(ggplot2)
source("../R/functions_da.R")
```

The data comes from the article of Ravel et. al. 2013.

Reference:

Ravel, J., Brotman, R.M., Gajer, P. et al. Daily temporal dynamics of vaginal
microbiota before, during and after episodes of bacterial vaginosis.
Microbiome 1, 29 (2013). https://doi.org/10.1186/2049-2618-1-29

```{r}
se <- readRDS("Ravel_2013_16S_BV_se.rds")
se
```

## Exploring sample metadata

```{r}
sample_metadata <- se %>% 
    colData() %>% 
    as.data.frame() %>% 
    rownames_to_column(var = "Sample") %>% 
    as_tibble() %>% 
    mutate(
        subjectID = as.factor(subjectID),
        pH = as.double(pH),
        Nugent_Score = as.integer(Nugent_Score),
        ## Is the sample positive to either SBV or ABV using Amsel?
        amsel_pos = case_when(SBV == "1" | ABV == "1" ~ 1, TRUE ~ 0),
        ## Classification according to the Amsel and the article's criteria
        amsel_res = case_when(
            SBV == "1" ~ "SBV", ABV == "1" ~ "ABV", TRUE ~ "Neg"
        ),
        nugent_res = case_when(
            ## Classification based on Nugent scores
            Nugent_Score <= 3 ~ "Negative",
            Nugent_Score >=4 & Nugent_Score <= 6 ~ "Intermediate",
            Nugent_Score >= 7 ~ "Positive"
        ),
        nugent_res = factor(
          nugent_res, levels = c("Negative", "Intermediate", "Positive")
        )
    )
```

```{r}
glimpse(sample_metadata)
```

```{r}
sample_metadata %>% 
    count(subjectID, nugent_res, name = "count") %>% 
    filter(!is.na(nugent_res)) %>%
    ggplot(aes(subjectID, count)) +
    geom_col(aes(fill = nugent_res), position = "stack") +
    geom_text(
        aes(label = count, group = nugent_res), 
        position = position_stack(vjust = 0.5)
    ) +
    theme_bw() +
    theme(
        axis.text.x = element_text(angle = 45, hjust = 1)
    )
```

## Select subjects

Subjects with nine or more samples classified in both positive and negative BV
were selected.

```{r}
subject_neg_pos_counts <- sample_metadata %>% 
  filter(
      !is.na(nugent_res),
      nugent_res != "Intermediate",
  ) %>% 
  count(subjectID, nugent_res) %>% 
  pivot_wider(names_from = nugent_res, values_from = n) %>% 
  drop_na() %>% 
  filter(Negative >= 9, Positive >= 9)

subject_neg_pos_counts
```

```{r}
subject_neg_pos_counts %>% 
  summarise(across(c(Negative, Positive), sum))
```

```{r}
selected_sample_metadata <- sample_metadata %>% 
    filter(
        subjectID %in% subject_neg_pos_counts$subjectID,
        nugent_res != "Intermediate"
    )
selected_samples <- selected_sample_metadata$Sample
```

## Nugent score vs pH of selected subjects

There should be a clear positive correlation pattern between pH and Nugent
score.

```{r}
selected_sample_metadata %>% 
    mutate(pH = as.double(pH), Nugent_Score = as.integer(Nugent_Score)) %>% 
    suppressWarnings() %>% 
    filter(!is.na(pH), !is.na(Nugent_Score)) %>% 
    ggplot(aes(Nugent_Score, pH )) +
    geom_boxplot(aes(group = Nugent_Score, color = nugent_res)) +
    # geom_point(size = .25, position = "jitter", shape = 1) +
    geom_smooth() +
    scale_x_continuous(breaks = seq(0, 10, 1)) + 
    theme_bw()
```

## Summarize by genus

```{r}
se_genus <- mia::agglomerateByRank(se, rank = "genus")
se_genus
```

Get taxonomy table: 

```{r}
taxa_genus <- se_genus %>% 
    rowData() %>% 
    as.data.frame %>% 
    rownames_to_column(var = "tax_name") %>% 
    as_tibble()
```

Combine sample metadata, relative abundance, and taxonomy into a single tibble:

```{r}
tidy_table_genus <- assays(se_genus)$relative_abundance %>% 
    as.data.frame() %>% 
    rownames_to_column(var = "Taxa") %>% 
    as_tibble() %>% 
    pivot_longer(names_to = "Sample", values_to = "Abundance", 2:last_col()) %>% 
    right_join(selected_sample_metadata, by = "Sample") %>% 
    right_join(taxa_genus, by = c("Taxa" = "tax_name")) %>% 
    relocate(Sample, Taxa, Abundance)
```


Nugent score vs relative abundance of Lactubacillus


```{r}
taxa_genera <- c(
    "Lactobacillus$",
    "Gardnerella$"
)
taxa_regex <- paste0("(", paste0(taxa_genera, collapse = "|"), ")")

tidy_table_genus %>% 
    filter(grepl(taxa_regex, Taxa)) %>%
    select(Sample, Taxa, Abundance, pH, Nugent_Score, nugent_res) %>% 
    mutate(
        Taxa = sub("^.+_", "", Taxa),
        Taxa = factor(Taxa, levels = sub("\\$$", "", taxa_genera)),
    ) %>% 
    drop_na() %>% 
    suppressWarnings() %>% 
    ggplot(aes(Nugent_Score, Abundance)) +
    geom_boxplot(
        aes(group = Nugent_Score, color = nugent_res), outlier.shape = 4
    ) +
    # geom_point(size = .25, position = "jitter") +
    geom_smooth(color = "red") +
    facet_wrap(~Taxa, ncol = 2) +
    scale_x_continuous(breaks = seq(0,10,1)) +
    scale_color_discrete(name = "BV diagnostic (Nugent score)") +
    labs(
        y = "Relative abundance", x = "Nugent score",
        title = "Ravel et. al., 2013"
    ) +
    theme_bw() +
    theme(
        legend.position = "bottom"
    )
```

```{r}
selected_se <- se[,selected_samples]
taxa_lgl <- rowSums(assays(selected_se)$counts) > 0
selected_se <- selected_se[taxa_lgl, ]
nugent_scores <- colData(selected_se)$Nugent_Score 
colData(selected_se)$BV <- 
  ifelse(nugent_scores >= 7, "Positive BV", "Negative BV")
table(colData(selected_se)$BV)
```

## Differential expression analysis

Some variables

```{r}
mode(assays(selected_se)$counts) <- "integer"
grp <- "BV"
ref <- "Negative BV"
```

```{r}
deseq_results <- deseq2Poscounts(selected_se, grp = grp, ref = ref)
```



```{r}
deseq_results %>% 
  filter(ADJPVAL <=0.1) 
```



