---
title: "Ravel_2011_16S - Bacterial vaginosis II"
output:
    html_document:
        toc: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, message=FALSE}
library(MicrobiomeBenchmarkData)
library(MicrobiomeBenchmarkDataAnalyses)
library(mia)
library(phyloseq)
library(dplyr)
library(benchdamic)


library(purrr)
library(ggplot2)
library(gridExtra)
```

# Import data

```{r}
dat_name <- 'Ravel_2011_16S_BV'
conditions_col <- 'study_condition'
conditions <- c(condB = 'healthy', condA = 'bacterial_vaginosis')

tse <- getBenchmarkData(dat_name, dryrun = FALSE)[[1]]

################################################################
## Select equal number of samples per ethnic group
col_data <- as.data.frame(colData(tse))

row_names <- col_data %>% 
    split(factor(.$ethnicity)) %>% 
    map(~split(., .x$study_condition)) %>% 
    unlist(recursive = FALSE) %>% 
    map(rownames)

min_n <- row_names %>% 
    map_int(length) %>% 
    min()

set.seed(4567)
select_samples <- map(row_names, ~ sample(.x, min_n, replace = FALSE)) %>% 
    unlist(use.names = FALSE)

##############################################
# aelect_samples <- colData(tse)[[conditions_col]] %in% conditions
tse <- tse[, select_samples]
tse <- tse[rowSums(assay(tse)) > 1,] 
tse
```

## Get prior Info by genus 

```{r}
prior_info <- as.data.frame(rowData(tse)[, c('genus', 'taxon_annotation')])
newNames <- paste0(rownames(prior_info), '|', prior_info$genus)
prior_info <- prior_info %>% 
    filter(!is.na(genus)) %>% 
    distinct()
rownames(prior_info) <- prior_info$genus
head(prior_info)
```

## Summarize by genus

```{r}
tse_genus <- splitByRanks(tse)$genus
tse_genus
```

Reorder prior_info according to rownames of se_genus

```{r}
prior_info <- prior_info[rownames(tse_genus),]
```

Study condition as factor with the levesl for healthy vagina and BV in the
right order

```{r}
ps <- makePhyloseqFromTreeSummarizedExperiment(tse_genus)
sample_data(ps)[[conditions_col]] <- 
    factor(sample_data(ps)[[conditions_col]], levels = conditions)

colData(tse)[[conditions_col]] <- 
    factor(colData(tse)[[conditions_col]], levels = conditions)

```


## Run Differential abundance analysis

Filter

```{r}
ps <- filter_phyloseq(ps)

tse_genus <- mia::makeTreeSummarizedExperimentFromPhyloseq(ps)
```

DA analysis:

```{r, message=FALSE}




ps <- runNormalizations(set_norm_list(),ps, verbose = FALSE)
# zw <- weights_ZINB(ps, design = conditions_col)
DA_methods <- set_DA_methods_list(conditions_col, conditions)

DA_methods$DA_ALDEx2.1 <- NULL
DA_methods <- purrr::discard(DA_methods, is.null)
```

```{r}
fname <- paste0('./', dat_name, '.rds')
fname2 <- paste0('../../misc/articles_data/', dat_name, '.rds')
if (file.exists(fname)) {
    message('DA output exists in current directory. Importing...')
    DA_output <- readRDS(fname)
} else if (file.exists(fname2)) {
    message('DA ouptut exists in misc directory. Importing...')
    DA_output <- readRDS(fname2)
} else {
    DA_output <- runDA(DA_methods, ps, weights = zw, verbose = FALSE) 
    saveRDS(DA_output, fname)
} 
```

See results

```{r}
remove_output <- c(
    "metagenomeSeq.CSSdefault", "wilcox.none", "ZINQ.none.MinP", 
    "ZINQ.TSS.MinP", "ZINQ.CLR.MinP", "ZINQ.none.Cauchy", "lefse.none"
)

DA_output <- DA_output[!names(DA_output) %in% remove_output]
names(DA_output)
```

# Enrichment

Get direction

```{r}
direction <- get_direction_cols(DA_output, conditions_col, conditions)
```


```{r}
enrichment <- createEnrichment(
    object = DA_output,
    priorKnowledge = prior_info,
    enrichmentCol = "taxon_annotation",
    namesCol = "genus",
    slot = "pValMat", colName = "adjP", type = "pvalue",
    direction = direction,
    threshold_pvalue = 0.1,
    threshold_logfc = 0,
    top = NULL, 
    alternative = "greater",
    verbose = FALSE 
)
```


```{r}
enrich_plot <- plot_enrichment(
    enrichment = enrichment, 
    enrichment_col = "taxon_annotation",
    levels_to_plot = c("hv-associated", "bv-associated"),
    conditions = c(condB = "HV", condA = "BV") 
)
enrich_plot
```

# Plot putative true positves and true negatives ratio

```{r}
positives <- createPositives(
    object = DA_output, 
    priorKnowledge = prior_info, 
    enrichmentCol = "taxon_annotation", namesCol = "genus",
    slot = "pValMat", colName = "adjP", type = "pvalue",
    direction = direction,
    threshold_pvalue = 1,
    threshold_logfc = 0,
    top = seq.int(from = 0, to = 20, by = 2),
    alternative = "greater",
    verbose = FALSE,
    TP = list(c("DOWN Abundant", "hv-associated"), c("UP Abundant", "bv-associated")),
    FP = list(c("DOWN Abundant", "bv-associated"), c("UP Abundant", "hv-associated"))
)
```


```{r}
meth_class <- get_meth_class() 
positives2 <- left_join(positives, meth_class, by = 'method') %>% 
    relocate(method_class)
```


```{r}
plots <- plot_positives(positives2)
grid.arrange(grobs = plots, ncol = 3)
```





```{r, eval=FALSE, include=FALSE, echo=FALSE}

ggsave(filename = "Ravell_2011_a.png", plot = enrich_plot, width = 8, height = 5)

positives_plots <- grid.arrange(grobs = plots, ncol = 3)
ggsave(filename = "Ravell_2011_b.png", plot = positives_plots, width = 10, height = 6)
```



