---
title: "Analysis of the HMP_2012_16S_gingival_V35_subset dataset"
output: 
    html_document:
        toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r, message=FALSE}
library(MicrobiomeBenchmarkData)
# library(MicrobiomeBenchmarkDataAnalyses)
library(phyloseq)
library(mia)
library(benchdamic)
library(purrr)
library(dplyr)
library(ggplot2)
library(gridExtra)
library(randomcoloR)
```

# Import Data

```{r}
tse <- getDataset("HMP_2012_16S_gingival_V35_subset", dryrun = FALSE)[[1]]
grp <- "hmp_body_subsite"
conditions <- c(condB = "subgingival_plaque", condA = "supragingival_plaque")
```

# Get prior knowledge (biological truth)

```{r}
row_data <- as.data.frame(rowData(tse))
taxaNames <- rownames(row_data)
newNames <- paste0(taxaNames, "|", row_data$GENUS) 
priorInfo <- data.frame(
    taxaNames = taxaNames,
    genus = row_data$GENUS,
    newNames = newNames,
    type = row_data$BIOSIS 
)
rownames(priorInfo) <- taxaNames
```

# Convert to phyloseq

```{r}
ps <- makePhyloseqFromTreeSummarizedExperiment(tse)

## This step places the first condition ('condB') as reference/denominator
sample_data(ps)[[grp]] <- factor(
    sample_data(ps)[[grp]], levels = conditions
)
```

# Run DA methods

```{r warning=FALSE, message=FALSE}
DA_output <- 
    run_DA(object = ps, conditions_col = grp, conditions = conditions)
```


# Enrichment

## Creaate enrichment object

## Plot enrichment


# Putative true positives - putative false positives

## Create enrichment object

## Plot enrichment



### LEFSE

```{r}
output_lefse_none <- DA_lefse(
    object = ps, grp = grp, ref = conditions[["condB"]],
    kruskal.threshold = 1,
    wilcox.threshold = 1,
    lda.threshold = 0,
    norm = 'none'
)

output_lefse_clr <- DA_lefse(
    object = ps, grp = grp, ref = conditions[["condB"]],
    kruskal.threshold = 1,
    wilcox.threshold = 1,
    lda.threshold = 0,
    norm = "CLR"
)
```


### Direction

```{r}
direction <- c(
    output_deseq2 = "log2FoldChange",
    output_deseq2_zinbweights = "log2FoldChange",
    output_edgeR = "logFC",
    output_edgeR_zinbweihgts = "logFC",
    output_limma = "logFC",
    output_limma_zinbweights = "logFC",
    output_aldex2_t = "effect",
    output_aldex2_wilcox = "effect",
    output_metagenomeseq_default = paste0(grp, conditions["condA"]),
    output_metagenomeseq_median = paste0(grp, conditions["condA"]),
    output_corncob = "Estimate",
    output_mast = "logFC",
    output_seurat = "avg_log2FC",
    output_wilcox_none = "log2FoldChange",
    output_wilcox_clr = "log2FoldChange",
    output_wilcox_tss = "log2FoldChange",
    output_zinq = "log2FoldChange",
    output_ancombc = "logFC",
    output_lefse_none = "LDA_scores",
    output_lefse_clr = "LDA_scores"
)

```








### Calculate enrichment

```{r}
enrichment <- createEnrichment(
    object = DA_results,
    priorKnowledge = priorInfo,
    enrichmentCol = "type",
    namesCol = "newNames",
    slot = "pValMat", colName = "rawP", type = "pvalue",
    direction = direction,
    threshold_pvalue = 0.05,
    threshold_logfc = 0,
    top = NULL,
    alternative = "greater",
    verbose = FALSE 
)

```


### Plot enrichment

```{r, warning=FALSE, message=FALSE, fig.height=6, fig.width=10}
enrich_plot <- plot_enrichment(
    enrichment = enrichment, 
    enrichment_col = "type",
    levels_to_plot = c("Aerobic", "Anaerobic", "F Anaerobic"),
    conditions = conditions
)
enrich_plot

```

## Putative positives ratio

### Create table of putative positives

```{r}
positives <- createPositives(
    object = DA_results, 
    priorKnowledge = priorInfo, enrichmentCol = "type", namesCol = "newNames",
    slot = "pValMat", colName = "rawP", type = "pvalue",
    direction = direction,
    threshold_pvalue = 0.05,
    threshold_logfc = 0,
    top = seq.int(from = 0, to = 50, by = 5),
    alternative = "greater",
    verbose = FALSE,
    TP = list(c("DOWN Abundant", "Anaerobic"), c("UP Abundant", "Aerobic")),
    FP = list(c("DOWN Abundant", "Aerobic"), c("UP Abundant", "Anaerobic"))
)
```


### Plot table of putative positives

```{r}
# pos_plot <- plotPositives(positives)
# pos_plot
```


```{r}
meth_class <- method_classification()
positives2 <- left_join(meth_class, positives, by = 'method') %>% 
    relocate(method_class)
# positives2 <- positives2 %>% 
#     mutate(
#         base_method = sub("\\..*", "", method)
#     )
```


```{r, fig.height=6, fig.width=10}

# n_meth <- length(unique(positives2$method))
# 
# p1 <- positives2 %>%    
#     mutate(
#         method = paste0(method_class, "--", method)
#     ) %>% 
#     ggplot(
#         aes(x = top, y = TP - FP)
#     ) +
#     geom_path(
#         aes(linetype = method_class, color = method)
#     ) +
#     geom_point(
#         aes(shape = method_class, color = method)
#     ) +
#     labs(x = "Top") +
#     scale_color_manual(values = distinctColorPalette(n_meth)) +
#     scale_linetype_discrete(guide = "none") +
#     scale_shape_discrete(guide = 'none') +
#     facet_wrap(~method_class) +
#     theme_bw() +
#     theme(
#         legend.title = element_blank()
#     )
# 
# p1

```



```{r}

max_top <- max(positives2$TP - positives2$FP)
min_top <- min(positives2$TP - positives2$FP)

x <- split(positives2, positives2$method_class)
df <- x[[2]]
df <- df %>% 
    mutate(
        across(.cols = where(is.character), .fns = ~ forcats::fct_inorder(.x)),
        shape = forcats::fct_inorder(as.character(shape))
        )
        

(
p1 <- df %>%
        ggplot2::ggplot(ggplot2::aes(x = top, y = TP - FP)) +
        ggplot2::geom_path(
            ggplot2::aes(color = color, group = method, linetype = norm)
        ) +
        ggplot2::geom_point(
            ggplot2::aes(color = color, shape = norm),
            size = 3
        ) +
        ggplot2::facet_wrap(.~ method_class) +
        ggplot2::scale_y_continuous(
            limits = c(min_top - 3, max_top + 3)
        ) +
        ggplot2::scale_color_manual(
            values = levels(df$color), labels = levels(df$base_method)
            
        ) +
        ggplot2::scale_shape_manual(
            values = as.integer(levels(df$shape)), labels = levels(df$norm)
        ) +
        ggplot2::scale_linetype_manual(
            values = as.integer(levels(df$shape)), labels = levels(df$norm)
        ) +
        ggplot2::labs(
            x = 'Top', y = 'TP - FP'
        ) +
        ggplot2::theme_bw() +
        ggplot2::theme(
            legend.position = c(0.01, 0.97),
            legend.justification = c("left", "top"),
            legend.title = ggplot2::element_blank(),
            legend.box.background = element_rect(color = "white", fill = 'white'
            )
        ) 
) 

```




```{r, fig.height = 12, fig.width=10}
plots <- plot_positives(positives2)

# first_plot <- plots[1] 
# other_plots <- plots[2:length(plots)]
# other_plots <- lapply(other_plots, function(x) {
#     x <- x +
#         theme(
#             axis.text.y = element_blank(),
#             axis.title.y = element_blank(),
#             axis.ticks.y = element_blank()
#         )
#     x
# })
# 
# plots2 <- c(first_plot, other_plots)

grid.arrange(grobs = plots, ncol = 2)


```


