---
title: "Ravel 2013 - Bacterial vaginosis"
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  message = FALSE,
  warning = FALSE
)
```

```{r setup}
library(dplyr)
library(SummarizedExperiment)
library(MicrobiomeBenchmarkDataAnalyses)
```

# Data

## Import data from the original article

Import OTU table (and a few metadata of the samples):

```{r}
## Data from supplementary table s03
otu_table_url <- "https://static-content.springer.com/esm/art%3A10.1186%2F2049-2618-1-29/MediaObjects/40168_2013_28_MOESM3_ESM.xlsx"
otu_table_file <- tempfile()
download.file(otu_table_url, otu_table_file)
otu_table <- readxl::read_xlsx(otu_table_file, range = "A4:EW1661")
head(otu_table)
```

Import sample metadata--some of the data is missing, but it's available
in the OTU table (above).

```{r}
## Data from supplementary table s01
sample_metadata_1_url <- "https://static-content.springer.com/esm/art%3A10.1186%2F2049-2618-1-29/MediaObjects/40168_2013_28_MOESM1_ESM.xlsx"
sample_metadata_1_file <- tempfile()
download.file(sample_metadata_1_url, sample_metadata_1_file)
sample_metadata_1 <- 
    readxl::read_xlsx(sample_metadata_1_file, range = "Metadata!A4:T1760")
head(sample_metadata_1)
```

Not all of the samples in the metadata are represented in the relative
abundance matrix. So the intersection is needed:

```{r}
samples <- intersect(otu_table$sampleID, sample_metadata_1$sampleID)
length(samples)
```

```{r}
sample_metadata <- sample_metadata_1 %>% 
    left_join(otu_table[c("sampleID", "Total Reads")], by = "sampleID") %>% 
    tibble::column_to_rownames(var = "sampleID") %>% 
    as.data.frame()
sample_metadata <- sample_metadata[samples, ]
sample_metadata <- sample_metadata %>% 
    mutate(
        Nugent_score_category = case_when(
            Nugent_Score >= 0 & Nugent_Score <= 3 ~ "low",
            Nugent_Score >= 4 & Nugent_Score <= 6 ~ "medium",
            Nugent_Score >= 7 & Nugent_Score >= 10 ~ "high"
        )
    )

head(sample_metadata)
```

```{r}
table(sample_metadata$Nugent_score_category)
```


```{r}
filtered_sample_metadata <- sample_metadata %>% 
    filter(Nugent_score_category %in% c("low", "high"))
table(filtered_sample_metadata$Nugent_score_category)
```

Extract the relative abundance matrix from the OTU table:

```{r}
relative_abundance_matrix <- otu_table %>% 
    select(-`Total Reads`) %>% 
    tibble::column_to_rownames(var = "sampleID") %>% 
    as.data.frame() %>% 
    as.matrix() %>% 
    t()
relative_abundance_matrix <- relative_abundance_matrix[, samples]
relative_abundance_matrix[1:5, 1:5]
```

```{r}
summary(colSums(relative_abundance_matrix))
```

Get counts for abundance matrix

```{r}
total_reads <- sample_metadata$`Total Reads`
abundance_matrix <- 
    relative_abundance_to_counts(relative_abundance_matrix, total_reads)
abundance_matrix[1:5, 1:5]
```

```{r, eval=FALSE, include=FALSE}
# abundance_matrix_fname <- 
#     "../../inst/extdata/Ravel_2013_bv_abundance_matrix.tsv"
# write.table(
#     abundance_matrix, abundance_matrix_fname,
#     sep = "\t", col.names = TRUE, row.names = TRUE
# )
```

Import taxonomy table

```{r}
taxonomy_table_file <- system.file(
    "extdata/Ravel_2013_bv_taxonomy_table.tsv",
    package = "MicrobiomeBenchmarkDataAnalyses"
)
taxonomy_table <- read.table(
    file = taxonomy_table_file, sep = "\t", header = TRUE, row.names = 1
)
head(taxonomy_table)
```

Create SummarizedExperiment

```{r}
filtered_samples <- rownames(filtered_sample_metadata)
filtered_abundance_matrix <- abundance_matrix[, filtered_samples]
filtered_relative_abundance_matrix <-
    relative_abundance_matrix[, filtered_samples]

se <- SummarizedExperiment(
    assays = S4Vectors::SimpleList(
        abundance = filtered_abundance_matrix, 
        relative_abundance = filtered_relative_abundance_matrix
    ),
    colData = S4Vectors::DataFrame(filtered_sample_metadata),
    rowData = S4Vectors::DataFrame(taxonomy_table)
)
se <- se[rowSums(assay(se)) > 0,]
se
```





