---
title: "Bacterial vaginosis dataset 1 - Ravel 2011"
output:
  html_document:
    toc: yes
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  message = FALSE
)
```

# Introduction

This is the analysis of the bacterial vaginosis dataset published
by Ravel, et. al., 2011 and available through a couple of sources:

+ The original data published by the authors on [PNAS](https://www.pnas.org/content/108/supplement_1/4680).
  + Supplementary table S4 (sample metadata and counts per taxa).
+ The data processed by the [Microbiome ML-repo](https://knights-lab.github.io/MLRepo/docs/ravel_nugent_category.html).

Packages:

```{r, message=FALSE}
library(dplyr)
library(tidyr)
library(ggplot2)
library(SummarizedExperiment)
library(mia)
library(purrr)
library(forcats)
library(phyloseq)
library(benchdamic)
library(MicrobiomeBenchmarkDataAnalyses)
```

# Data 

## Sample metadata

Sample metadata can be obtained combining data from the original article 
(Supplementary Table S04) and from the NCBI SRA (accession SRA022855).

Data from the NCBI SRA:

```{r}
## The file containing sample metadata for the SRA accession was downloaded
## with E-utilites from the NCBI
path_to_sra_metadata <- system.file(
  "extdata/SRA022855_run_metadata.csv",
  package = "MicrobiomeBenchmarkDataAnalyses"
)
sra_metadata <- read.csv(path_to_sra_metadata)
sra_metadata <- sra_metadata %>% 
  select(Run, SampleName, Platform, Model, CenterName)
head(sra_metadata)
```

Sample metadata provided by the authors:

```{r}
supp_tbl_s04_url <- "https://www.pnas.org/highwire/filestream/603190/field_highwire_adjunct_files/4/st04.xlsx"
supp_tbl_s04_file <- tempfile()
download.file(url = supp_tbl_s04_url, destfile = supp_tbl_s04_file)
supp_tbl_s04 <- readxl::read_xlsx(
  path = supp_tbl_s04_file, sheet = 1, range = "A3:IU397"
)
head(supp_tbl_s04)
```

Combine data from the authors and the SRA to create sample metadata:

```{r}
sample_metadata <- supp_tbl_s04[,1:7] %>% 
  left_join(sra_metadata, by = c("Subject ID" = "SampleName")) %>% 
  magrittr::set_colnames(sub("[a-d]$", "", colnames(.))) %>% 
  magrittr::set_colnames(gsub(" ", "", colnames(.))) %>% 
  as.data.frame()
rownames(sample_metadata) <- sample_metadata$Run
head(sample_metadata)
```

The sample metadata must be filtered with low and high Nugent_score categories.
This filtered sample metadata will be used to create the SummarizedExperiments
and phyloseq objects needed for analysis:

```{r}
filtered_sample_metadata <- sample_metadata %>% 
  filter(Nugentscorecategory %in% c("low", "high"))
table(filtered_sample_metadata$Nugentscorecategory)
```

A few plots. Nugent score vs Nugent category:

```{r}
filtered_sample_metadata %>% 
  ggplot(aes(Nugentscorecategory, Nugentscore)) +
  geom_boxplot() +
  labs(
    x = "Nugent score category", y = "Nugent score",
    title = "Nugent scores and category of the samples in Ravel 2011"
  ) +
  theme_bw()
  
```

Nugent category vs pH:

```{r}
filtered_sample_metadata %>% 
  ggplot(aes(Nugentscorecategory, pH)) +
  geom_violin() +
  labs(
    x = "Nugent category",
    title = "Nugent category and pH of the samples in Ravel 2011"
  ) +
  theme_bw()
```
 
Relationship between nugent scores and pH:
 
```{r}
filtered_sample_metadata %>% 
  ggplot(aes(Nugentscore, pH)) +
  geom_point(position = "jitter") +
  labs(
    x = "Nugent score",
    title = "Nugent score vs pH of the samples in Ravel 2011"
  ) +
  theme_bw()
```

## Abundance data

There are three main datasets with abundance data:
+ Data from the orginal article.
+ Data generated from the gg97 database.
+ Data generated from the RefSeq database.

### Original article

#### Abundance matrix

Abundance data is also included in the Supplementary Table S04 of the original
article:

```{r}
relative_abundance_matrix <- dplyr::left_join(
 sample_metadata, supp_tbl_s04, by = c("SubjectID" = "Subject ID")
)
relative_abundance_matrix <- 
  dplyr::select(relative_abundance_matrix, Run, `L. iners`:last_col())
rownames(relative_abundance_matrix) <- relative_abundance_matrix[,1]
relative_abundance_matrix <- relative_abundance_matrix[,-1]
relative_abundance_matrix <- t(relative_abundance_matrix)
rownames(relative_abundance_matrix) <- sub("^L\\.", "Lactobacillus", rownames(relative_abundance_matrix))
relative_abundance_matrix[1:5, 1:5]
```

The data is relative abundance:

```{r}
colSums(relative_abundance_matrix) %>% 
  summary()
```

```{r}
if (all(colnames(relative_abundance_matrix) == sample_metadata$Run)) {
  abundance_matrix <- apply(
    relative_abundance_matrix, 
    1, 
    function(x) x * sample_metadata$Totalnumberreads / 100
  ) %>% 
    t() %>% 
    round()
}
abundance_matrix[1:5, 1:5]
```

Just a quick comparison to be sure that the library size are approximately 
the same:

```{r}
all.equal(unname(colSums(abundance_matrix)), sample_metadata$Totalnumberreads)
```

```{r}
cbind(unname(colSums(abundance_matrix)), sample_metadata$Totalnumberreads) %>% 
  head()
```

```{r}
cbind(unname(colSums(abundance_matrix)), sample_metadata$Totalnumberreads) %>% 
  tail()
```

```{r}
summary(colSums(abundance_matrix) -sample_metadata$Totalnumberreads)
```

Abundance matrix of the original paper is ready.

```{r save abundance matrix, eval=FALSE, include=FALSE}
## Save the abundance matrix
# write.table(
#   abundance_matrix,
#   file = "/home/samuelgamboa/Projects/MicrobiomeBenchmarkDataAnalyses/MicrobiomeBenchmarkDataAnalyses/inst/extdata/Ravel_2011_bv_abundance_matrix.tsv",
#   sep = "\t", row.names = TRUE, col.names = TRUE
# )
```

#### Taxonomic table

The taxonomy table has been created beforehand (this is for convenience) with
the `taxize` package. So it only needs to be imported.

```{r}
fname <- system.file(
  "extdata/Ravel_2011_bv_taxonomy_table.tsv",
  package = "MicrobiomeBenchmarkDataAnalyses"
)
taxonomy_table <- read.table(fname, sep = "\t", row.names = 1, header = TRUE)
head(taxonomy_table)
```

#### SummarizedExperiment

We can create a SummarizedExperiment with the data above:

```{r}
assay_matrix <- abundance_matrix[, rownames(filtered_sample_metadata)]
se <- SummarizedExperiment(
  assays = S4Vectors::SimpleList(abundance = assay_matrix),
  rowData = S4Vectors::DataFrame(taxonomy_table),
  colData = S4Vectors::DataFrame(filtered_sample_metadata)
)
## Since a the samples were filtered is necesary to remove taxa
## tha have 0 values in the rowSums.
se <- se[rowSums(assay(se)) > 0,]
se
```

Create a SummarizedExperiment at the genus level (I think this is the best
option to use this dataset):

```{r}
se_ranks <- splitByRanks(se)
se_genus <- se_ranks$genus
assays(se_genus)$relative_abundance <- 
  apply(assay(se_genus), 2, function(x) x / sum(x) * 100)
se_genus
```

### gg97 (Microbiome ML-repo) 

#### OTU table

```{r}
otu_table_gg97_url <- 
    "https://knights-lab.github.io/MLRepo/datasets/ravel/gg/otutable.txt"
otu_table_gg97 <- 
  read.table(otu_table_gg97_url, sep = "\t", header = TRUE, comment.char = "")
rownames(otu_table_gg97) <- otu_table_gg97[[1]]
otu_table_gg97 <- as.matrix(otu_table_gg97[,-1])
otu_table_gg97[1:5, 1:5]
```

#### Taxa table

```{r}
taxa_table_gg97_url <- 
    "https://knights-lab.github.io/MLRepo/datasets/ravel/gg/taxatable.txt"
taxa_table_gg97 <- 
  read.table(taxa_table_gg97_url, sep = "\t", header = FALSE)
rownames(taxa_table_gg97) <- taxa_table_gg97[,1]
taxa_table_gg97 <- as.matrix(taxa_table_gg97[,-1])
taxa_table_gg97[1:5, 1:5]
```

Let's add some colnames. The OTU table and the taxa table contain the same
number of reads per sample. The difference is that the OTU table has all
individual OTUs whereas the taxa table contains the abundance values summarized
by taxa (different levels).

```{r}
if (all(colSums(otu_table_gg97) == colSums(taxa_table_gg97))) {
    abundance_matrix_gg97 <- taxa_table_gg97
    colnames(abundance_matrix_gg97) <- colnames(otu_table_gg97)
}
colnames(abundance_matrix_gg97) %>% 
    head()
```

The data above will be used for analysis.

#### Taxonomic classification

```{r}
taxonomic_table_gg97 <- rownames(abundance_matrix_gg97) %>% 
    as.data.frame() %>% 
    magrittr::set_colnames("taxa") %>% 
    mutate(taxa = sub(" p__\\w+;", "", taxa)) %>% 
    separate(
        col = "taxa", 
        into = c("kingdom", "class", "order", "family", "genus", "species"),
        sep = ";",
        remove = TRUE
    ) %>% 
    mutate(
        across(.cols = everything(), .fns = ~ sub("\\s*\\w__", "", .x)),
        species = ifelse(!is.na(species), paste0(genus, "_", species), species)
    )
rownames(taxonomic_table_gg97) <- rownames(abundance_matrix_gg97)
head(taxonomic_table_gg97)
```

#### SummarizedExperiment

```{r}
assay_matrix_gg97 <- abundance_matrix_gg97[, rownames(filtered_sample_metadata)]
se_gg97 <- SummarizedExperiment(
  assays = S4Vectors::SimpleList(abundance = assay_matrix_gg97),
  rowData = S4Vectors::DataFrame(taxonomic_table_gg97),
  colData = S4Vectors::DataFrame(filtered_sample_metadata)
)
se_gg97 <- se_gg97[rowSums(assay(se_gg97)) > 0,]
se_gg97
```

Create a Summarized Experiment at the Genus level

```{r}
se_gg97_ranks <- splitByRanks(se_gg97)
se_gg97_genus <- se_gg97_ranks$genus
assays(se_gg97_genus)$relative_abundance <- 
  apply(assay(se_gg97_genus), 2, function(x) x / sum(x) * 100)
se_gg97_genus
```

### RefSeq (Microbiome ML-Repo)

#### OTU table

```{r}
otu_table_refseq_url <- 
    "https://knights-lab.github.io/MLRepo/datasets/ravel/refseq/otutable.txt"
otu_table_refseq <- 
  read.table(otu_table_refseq_url, sep = "\t", header = TRUE, comment.char = "")
rownames(otu_table_refseq) <- otu_table_refseq[[1]]
otu_table_refseq <- as.matrix(otu_table_refseq[,-1])
otu_table_refseq[1:5, 1:5]
```

#### Taxa table

```{r}
taxa_table_refseq_url <- 
    "https://knights-lab.github.io/MLRepo/datasets/ravel/refseq/taxatable.txt"
taxa_table_refseq <- 
  read.table(taxa_table_refseq_url, sep = "\t", header = FALSE)
rownames(taxa_table_refseq) <- taxa_table_refseq[,1]
rownames(taxa_table_refseq)[7] <- 
  "k__unclassified;c__unclassified;o__unclassified;f__unclassified;g__unclassified;s__unclassified;t__unclassified"
taxa_table_refseq <- as.matrix(taxa_table_refseq[,-1])
taxa_table_refseq[1:5, 1:5]
```

Let's add some colnames

```{r}
if (all(colSums(otu_table_refseq) == colSums(taxa_table_refseq))) {
    abundance_matrix_refseq <- taxa_table_refseq
    colnames(abundance_matrix_refseq) <- colnames(otu_table_refseq)
}
colnames(abundance_matrix_refseq) %>% 
    head()
```

The data above will be used for analysis.

#### Taxonomic classification

```{r}
taxonomic_table_refseq <- rownames(abundance_matrix_refseq) %>% 
    as.data.frame() %>% 
    magrittr::set_colnames("taxa") %>% 
    mutate(taxa = sub("p__\\w+;", "", taxa)) %>% 
    separate(
        col = "taxa", 
        into = c("kingdom", "class", "order", "family",
                 "genus", "species", "strain"),
        sep = ";",
        remove = TRUE
    ) %>% 
    mutate(
        across(.cols = everything(), .fns = ~ sub("\\s*\\w__", "", .x)),
        across(.cols = everything(), .fns = ~ gsub(" ", "_", .x))
    )
rownames(taxonomic_table_refseq) <- rownames(abundance_matrix_refseq)
head(taxonomic_table_refseq)
```

#### SummarizedExperiment

```{r}
assay_matrix_refseq <- abundance_matrix_refseq[, rownames(filtered_sample_metadata)]
se_refseq <- SummarizedExperiment(
  assays = S4Vectors::SimpleList(abundance = assay_matrix_refseq),
  rowData = S4Vectors::DataFrame(taxonomic_table_refseq),
  colData = S4Vectors::DataFrame(filtered_sample_metadata)
)
se_refseq <- se_refseq[rowSums(assay(se_refseq)) > 0,]
se_refseq
```

Let's create a SummarizedExperiment at the genus level:

```{r}
se_refseq_ranks <- splitByRanks(se_refseq)
se_refseq_genus <- se_refseq_ranks$genus
assays(se_refseq_genus)$relative_abundance <- 
  apply(assay(se_refseq_genus), 2, function(x) x / sum(x) * 100)
se_refseq_genus
```

# Comparison of the three datasets

```{r}
datasets <- 
  list(original = se_genus, gg97 = se_gg97_genus, refseq = se_refseq_genus)
map(datasets, dim)
```

```{r}
x <- filtered_sample_metadata[,c("Run", "Totalnumberreads")]
y <- datasets %>% 
  map(~ colSums(assay(.x))) %>% 
  reduce(bind_cols) %>% 
  magrittr::set_colnames(c("original", "gg97", "refseq")) %>% 
  bind_cols(x, .)
head(y)
```

# Exploratory analysis

## Original article

Create a tidy dataframe:

```{r}
mat <- assays(se_genus)$relative_abundance
abundance_df <- mat %>% 
  as_tibble(rownames = "Taxa") %>% 
  pivot_longer(names_to = "Sample", values_to = "Relative abundance", 2:last_col()) %>% 
  left_join(filtered_sample_metadata, by = c("Sample" = "Run"))
head(abundance_df)
```

Who are the top taxa in samples with low and high Nugent scores:

```{r}
summary_stats <- abundance_df %>% 
    group_by(Taxa, Nugentscorecategory) %>% 
    summarise(
        mean = mean(`Relative abundance`, na.rm = TRUE)
        ) %>% 
    ungroup()

summary_stats_divided <- 
    split(summary_stats, factor(summary_stats$Nugentscorecategory))

top_genera <- summary_stats_divided %>% 
  map(~ head(arrange(.x, desc(mean)), 7)) %>% 
  bind_rows(.id = "nugent_score_category")

top_genera
```

Create a plot:

```{r}
genera <- c(
  "Lactobacillus", "Prevotella", "Atopobium", "Megasphaera",
  "Sneathia", "Dialister"
)
abundance_df %>% 
  filter(Taxa %in% genera) %>% 
  mutate(
    Nugentscorecategory = fct_relevel(Nugentscorecategory, "low"),
    Taxa = fct_relevel(Taxa, "Lactobacillus", "Prevotella")
  ) %>% 
  ggplot(aes(Nugentscorecategory, `Relative abundance`)) +
  geom_boxplot() +
  facet_wrap(~Taxa) +
  labs(
    x = "Nugent score category",
    title = "Main taxa affected by BV (high Nugent score) - Original article"
  ) +
  theme_bw()
```

## gg97

Create a tidy dataframe:

```{r}
mat_gg97 <- assays(se_gg97_genus)$relative_abundance
abundance_df_gg97 <- mat_gg97 %>% 
  as_tibble(rownames = "Taxa") %>% 
  pivot_longer(names_to = "Sample", values_to = "Relative abundance", 2:last_col()) %>% 
  left_join(filtered_sample_metadata, by = c("Sample" = "Run"))
head(abundance_df_gg97)
```

Who are the top taxa in samples with low and high Nugent scores:

```{r}
summary_stats_gg97 <- abundance_df_gg97 %>% 
    group_by(Taxa, Nugentscorecategory) %>% 
    summarise(
        mean = mean(`Relative abundance`, na.rm = TRUE)
        ) %>% 
    ungroup()

summary_stats_divided_gg97 <- 
    split(summary_stats_gg97, factor(summary_stats_gg97$Nugentscorecategory))

top_genera_gg97 <- summary_stats_divided_gg97 %>% 
  map(~ head(arrange(.x, desc(mean)), 7)) %>% 
  bind_rows(.id = "nugent_score_category")

top_genera_gg97
```

Create a plot:

```{r}
genera_gg97 <- c(
  "Lactobacillus", "Prevotella", "Atopobium", "Megasphaera",
  "Sneathia", "Shuttleworthia"
)
abundance_df_gg97 %>% 
  filter(Taxa %in% genera_gg97) %>% 
  mutate(
    Nugentscorecategory = fct_relevel(Nugentscorecategory, "low"),
    Taxa = fct_relevel(Taxa, "Lactobacillus", "Prevotella")
  ) %>% 
  ggplot(aes(Nugentscorecategory, `Relative abundance`)) +
  geom_boxplot() +
  facet_wrap(~Taxa) +
  labs(
    x = "Nugent score category",
    title = "Main taxa affected by BV (high Nugent score) - gg97"
  ) +
  theme_bw()
```

## RefSeq

Create a tidy dataframe:

```{r}
mat_refseq <- assays(se_refseq_genus)$relative_abundance
abundance_df_refseq <- mat_refseq %>% 
  as_tibble(rownames = "Taxa") %>% 
  pivot_longer(names_to = "Sample", values_to = "Relative abundance", 2:last_col()) %>% 
  left_join(filtered_sample_metadata, by = c("Sample" = "Run"))
head(abundance_df_refseq)
```

Who are the top taxa in samples with low and high Nugent scores?

```{r}
summary_stats_refseq <- abundance_df_refseq %>% 
    group_by(Taxa, Nugentscorecategory) %>% 
    summarise(
        mean = mean(`Relative abundance`, na.rm = TRUE)
        ) %>% 
    ungroup()

summary_stats_divided_refseq <- 
    split(summary_stats_refseq, factor(summary_stats_refseq$Nugentscorecategory))

top_genera_refseq <- summary_stats_divided_refseq %>% 
  map(~ head(arrange(.x, desc(mean)), 7)) %>% 
  bind_rows(.id = "nugent_score_category")

top_genera_refseq
```

Create a plot:

```{r}
genera_refseq <- c(
  "Lactobacillus", "Prevotella", "Atopobium",
  "Sneathia", "Peptoniphilus", "Aerococcus"
)
abundance_df_refseq %>% 
  filter(Taxa %in% genera) %>% 
  mutate(
    Nugentscorecategory = fct_relevel(Nugentscorecategory, "low"),
    Taxa = fct_relevel(Taxa, "Lactobacillus", "Prevotella")
  ) %>% 
  ggplot(aes(Nugentscorecategory, `Relative abundance`)) +
  geom_boxplot() +
  facet_wrap(~Taxa) +
  labs(
    x = "Nugent score category",
    title = "Main taxa affected by BV (high Nugent score) - RefSeq"
  ) +
  theme_bw()
```

# Benchmarking of DA methods

## Preparation

### Set up of variables for performance of DA methods

The following variables can be used for the analysis of the three
different versions of the Ravel_2011 dataset:

Column indicating the labels of the samples:

```{r}
## Name of the column with the conditions to be compared.
## This column will be added to colData in the code below
grp <- "grp"

## Name of the conditions to be compared.
## Reference should be first always.
contrast <- c("low", "high")
```

### Set up prior biologcial info for enrichment

In this case we'll use three different datasets:
+ aerophilicity from the `waldronlab/bugphyzz` package.
+ gram staining from the `waldronlab/bugphyzz` package.
+ aerohilicyt from the biosis dataset of the `waldronlab/nychanes` package.

Import the aerophilicity dataset from bugphyzz

```{r}
fname1 <- system.file(
  "extdata/bugphyzz_aerophilicity.tsv", package = "MicrobiomeBenchmarkDataAnalyses"
)
aer <- readr::read_tsv(fname1) %>% 
  magrittr::set_colnames(c("Taxon_name", "aerophilicity"))

## Is necessary no remove the genus Haemophilus with aerobic annotation
## since this organism is also annotated as facultatively anaerobic. The latter
## coincides with the biosis dataset of the nychanes package.
aer <- aer %>% 
  filter(!(Taxon_name == "Haemophilus" & aerophilicity == "aerobic"))

head(aer)
```

Import the gram stain dataset from bugphyzz

```{r}
fname2 <- system.file(
  "extdata/bugphyzz_gram_stain.tsv", package = "MicrobiomeBenchmarkDataAnalyses"
)
gs <- readr::read_tsv(fname2) %>% 
  magrittr::set_colnames(c("Taxon_name", "gram_stain"))
head(gs)
```

Import the aerophilicity data from nychanes (biosis):

```{r}
fname3 <- system.file(
  "extdata/nychanes_biosis.tsv", package = "MicrobiomeBenchmarkDataAnalyses"
)
biosis <- readr::read_tsv(fname3)
head(biosis)
```

## Original article

### Run DA methods

Create phyloseq object:

```{r}
colData(se_genus)$grp <- 
  factor(colData(se_genus)$Nugentscorecategory, levels = contrast)
ps_genus <- makePhyloseqFromTreeSummarizedExperiment(
  se_genus, abund_values = "abundance"
)
ps_genus <- filter_phyloseq(ps_genus)
ps_genus
```

Set reference:

```{r}
## Using "low" as reference (contast[1])
sample_data(ps_genus)[[grp]] <- factor(sample_data(ps_genus)[[grp]])
sample_data(ps_genus)[[grp]] <- 
  relevel(x = sample_data(ps_genus)[[grp]], ref = contrast[1])
```

Estimate normalization factors:

```{r}
ps_genus <- runNormalizations(
  normalization_list = normalization_methods_list, object = ps_genus
) %>% 
  quiet()

## Calculate weights
zinbweights <- weights_ZINB(
    object = ps_genus,
    K = 0,
    design = "~ 1"
)

## Run DA methods
DA_methods_list <- set_DA_methods_list(
  grp = grp, contrast = contrast
)
DA <- runDA(
  method_list = DA_methods_list, object = ps_genus, weights = zinbweights
) %>% 
  quiet()

str(DA, max.level = 1)
```

## Perform enrichment

Extract table with taxonomy information:

```{r}
taxTable <- tax_table(ps_genus) %>% 
  as.data.frame() %>% 
  as_tibble(rownames = "Taxa")
head(taxTable)
```

Combine the taxonomy table with the data from bugphyzz and nychanes to
create the prior biologocial knowledge table:

```{r}
taxTable2 <- taxTable %>% 
  left_join(biosis, by = c("Taxa" = "genera")) %>% 
  left_join(aer, by = c("Taxa" = "Taxon_name")) %>% 
  left_join(gs, by = c("Taxa" = "Taxon_name")) %>% 
  as.data.frame()
rownames(taxTable2) <- taxTable2$Taxa

head(taxTable2)
```

Prepare direction for entichment:

```{r}
direction <- c(
    edgeR.TMM = "logFC", 
    DESeq2.poscounts = "log2FoldChange",
    limma.CSSmedian = "logFC",
    limma.TMM = "logFC",
    metgenomeSeq.CSSmedian = paste0(grp, contrast[2]),
    ALDEx2.none = "effect",
    corncob.none = "Estimate",
    MAST.none = "logFC",
    Seurat.none = "avg_log2FC",
    ZINQ = "log2FoldChange",
    ANCOMBC = "logFC",
    wilcox_test = "log2FoldChange",
    wilcox_test_clr = "log2FoldChange"
)
```

### Gram stain enrichment (bugphyzz)

Run enrichment (gram stain):

```{r}
enrichment_gs <- createEnrichment(
    object = DA,
    priorKnowledge = taxTable2,
    enrichmentCol = "gram_stain",
    namesCol = "Taxa",
    slot = "pValMat",
    colName = "adjP",
    type = "pvalue",
    direction = direction,
    threshold_pvalue = 0.1,
    threshold_logfc = 0,
    top = NULL,
    alternative = "greater",
    verbose = TRUE
)
```

Plot enrichment:

```{r}
plotEnrichment(
    enrichment = enrichment_gs, 
    enrichmentCol = "gram_stain",
    levels_to_plot = c("gram stain positive", "gram stain negative")
)
```

Create table of positives:

```{r}
positives_gs <- createPositives(
    object = DA, 
    priorKnowledge = taxTable2,
    enrichmentCol = "gram_stain",
    namesCol = "Taxa",
    slot = "pValMat",
    colName = "rawP",
    type = "pvalue",
    direction = direction,
    threshold_pvalue = 1,
    threshold_logfc = 0,
    top = seq.int(from = 0, to = 50, by = 5),
    alternative = "greater",
    verbose = FALSE,
    TP = list(c("DOWN Abundant", "gram stain positive"), c("UP Abundant", "gram stain negative")),
    FP = list(c("DOWN Abundant", "gram stain negative"), c("UP Abundant", "gram stain positive"))
)
head(positives_gs)
```

Plot positives

```{r}
plotPositives(positives_gs)
```

### aerophilicity enrichment (bugphyzz)

```{r}
enrichment_aer <- createEnrichment(
    object = DA,
    priorKnowledge = taxTable2,
    enrichmentCol = "aerophilicity",
    namesCol = "Taxa",
    slot = "pValMat",
    colName = "adjP",
    type = "pvalue",
    direction = direction,
    threshold_pvalue = 0.1,
    threshold_logfc = 0,
    top = NULL,
    alternative = "greater",
    verbose = TRUE
)
```

Plot enrichment:

```{r}
plotEnrichment(
    enrichment = enrichment_aer, 
    enrichmentCol = "aerophilicity",
    levels_to_plot = c("aerobic", "anaerobic", "facultatively anaerobic")
)
```


Create table of positives

```{r}
positives_aer <- createPositives(
    object = DA, 
    priorKnowledge = taxTable2,
    enrichmentCol = "aerophilicity",
    namesCol = "Taxa",
    slot = "pValMat",
    colName = "rawP",
    type = "pvalue",
    direction = direction,
    threshold_pvalue = 1,
    threshold_logfc = 0,
    top = seq.int(from = 0, to = 50, by = 5),
    alternative = "greater",
    verbose = FALSE,
    TP = list(c("DOWN Abundant", "aerobic"), c("UP Abundant", "anaerobic")),
    FP = list(c("DOWN Abundant", "anaerobic"), c("UP Abundant", "aerobic"))
)
head(positives_aer)
```

Plot positives

```{r}
plotPositives(positives_aer)
```


### aerophilicity enrichment (nychanes)

```{r}
enrichment_biosis <- createEnrichment(
    object = DA,
    priorKnowledge = taxTable2,
    enrichmentCol = "biosis",
    namesCol = "Taxa",
    slot = "pValMat",
    colName = "adjP",
    type = "pvalue",
    direction = direction,
    threshold_pvalue = 0.1,
    threshold_logfc = 0,
    top = NULL,
    alternative = "greater",
    verbose = TRUE
)
```

Plot enrichment:

```{r}
plotEnrichment(
    enrichment = enrichment_biosis, 
    enrichmentCol = "biosis",
    levels_to_plot = c("Aerobic", "Anaerobic", "F Anaerobic")
)
```


Create table of TRUE and FALSE positives:

```{r}
positives_biosis <- createPositives(
    object = DA, 
    priorKnowledge = taxTable2,
    enrichmentCol = "biosis",
    namesCol = "Taxa",
    slot = "pValMat",
    colName = "rawP",
    type = "pvalue",
    direction = direction,
    threshold_pvalue = 1,
    threshold_logfc = 0,
    top = seq.int(from = 0, to = 50, by = 5),
    alternative = "greater",
    verbose = FALSE,
    TP = list(c("DOWN Abundant", "Aerobic"), c("UP Abundant", "Anaerobic")),
    FP = list(c("DOWN Abundant", "Anaerobic"), c("UP Abundant", "Aerobic"))
)
head(positives_biosis)
```

Plot positives:

```{r}
plotPositives(positives_biosis)
```






## gg97


## Refseq
